### PR TITLE
Dataloader with dicts

### PR DESCRIPTION
**Pull request by AlexGrig**
If a DataLoader you are working with does not provide outputs of (Xs,Ys) but instead output dict with values then this PR allows to handle this. It is possible to redefine the DataLoader itself, but it is not always convenient. I have added an extra parameter to range_test with the class name. This class is supposed to be inherited from  DataLoaderIterWrapper. In the new calss redefine method _batch_make_inputs_labels of the new class so that it return tuple (Xs, Ys). There is also an example of usage.

### CODE REVIEW

**Comment by NaleRaphael** [comment]
Hi @AlexGrig, thanks for the contribution.

For this case, I would consider replacing the following line with a method as an overridable interface for users:
https://github.com/davidtvs/pytorch-lr-finder/blob/b0b0ecd28c0578ad1d5ed06d2442f9a2c29c72e2/torch_lr_finder/lr_finder.py#L189

(to something like this)
```python
class LRFinder(object):
    def range_test(...):
        # ...

        # Create an iterator to get data batch by batch
        iter_wrapper = self.make_iterator(train_loader)

        # ...

    def make_iterator(self, dataloader):
        # Override this method to use a custom `DataLoaderIterWrapper`
        return DataLoaderIterWrapper(train_loader)
```


With this approach, users can handle their complex dataset and make it able to conform with the format we used. This implementation should be more convenient and make less changes to the original codebase.
```python
class CustomDataloaderIterWrapper(object):
    def __next__(self):
        # ...
        return inputs, labels


class CustomLRFinder(LRFinder):
    def make_iterator(self, dataloader):
        return CustomDataloaderIterWrapper(dataloader)
```

And I think it would be better to wait for @davidtvs's response before making further changes.

**Comment by davidtvs** [comment]
Thanks for the PR @AlexGrig.

I would actually suggest something a mix of ideas from @AlexGrig and @NaleRaphael. My suggestion is to create a base class from the current `DataLoaderIterWrapper`:
```py
class DataLoaderIter(object):
    def __init__(self, data_loader, auto_reset=True):
        self.data_loader = data_loader
        self.auto_reset = auto_reset
        self._iterator = iter(data_loader)

    def inputs_labels_from_batch(self, batch_data):
        inputs, labels, *_ = batch_data

        return inputs, labels

    def __next__(self):
        return self.inputs_labels_from_batch(batch)
```

The current `DataLoaderIterWrapper` becomes a child class that overrides the `__next__` method:
```py
class TrainDataLoaderIter(DataLoaderIter):
    def __next__(self):
        try:
            batch = next(self._iterator)
            inputs, labels = self.inputs_labels_from_batch(batch)
        except StopIteration:
            if not self.auto_reset:
                raise
            self._iterator = iter(self.data_loader)
            batch = next(self._iterator)
            inputs, labels = self.inputs_labels_from_batch(batch)

        return inputs, labels
```

The user can then create its own custom dataloader wrapper in a similar fashion:
```py
class CustomTrainDataLoaderIter(TrainDataLoaderIter):
    def inputs_labels_from_batch(self, batch_data):
        # Manipulate the batch_data and to get inputs and labels
        # ...

        return inputs, labels
```

I've named these classes with the word `Train` because we actually have the same problem for the validation dataloader. It currently assumes that it returns a pair of `inputs, labels` but that might not be true for all users. So I suggest creating a `ValDataLoaderIter`
```py
class ValDataLoaderIter(DataLoaderIter):
    pass
```

which can also be overridden like this:
```py
class CustomValDataLoaderIter(ValDataLoaderIter):
    def inputs_labels_from_batch(self, batch_data):
        # Manipulate the batch_data and to get inputs and labels
        # ...
        return inputs, labels
```

The only thing left to do is integrate this into the `range_test` method. For this, and to avoid API changed relative to older versions of this package I propose that the parameters `train_loader` and `val_loader` can either be instances of the  `DataLoader` class or of the `TrainDataLoaderIter` and `ValDataLoaderIter` classes, respectively, and we would handle it like so:

```py
class LRFinder(object):
    def range_test(...):
        # ...
        if isinstance(train_loader, DataLoader):
            train_iter = TrainDataLoaderIter(train_loader)
        else if isinstance(train_loader, TrainDataLoaderIter):
            train_iter = train_loader
        else:
            raise ValueError("...")
        
        if isinstance(val_loader, DataLoader):
            val_iter = ValDataLoaderIter(val_loader)
        else if isinstance(val_loader, ValDataLoaderIter):
            val_iter = val_loader
        else:
            raise ValueError("...")
```

Further changes would be needed within `range_test`, `_train_batch`, and `_validate`. I would like a better name to replace `train_loader` and `val_loader` but currently can't think of one.

What do you guys think?

**Comment by NaleRaphael** [comment]
@davidtvs Great idea, I second this approach.

> I would like a better name to replace `train_loader` and `val_loader` but currently can't think of one.

If I understand correctly, the idea of changing these names is to make them represent their own types unambiguously, i.e. they are `DataLoaderIter` instead of `DataLoader`. 

In my opinion, it's fine to keep them unchanged since users usually pass `DataLoader` to it. And we can revise those parts related to `train_loader` and `val_loader` in the docstring to make users know that `DataLoaderIter` is also an acceptable type of argument. Add some examples of using `DataLoaderIter` for this case in README.md might be helpful, too.

```python
def range_test(...):
    """
    Arguments:
        train_loader (torch.utils.data.DataLoader or DataLoaderIter): ...
        val_loader (torch.utils.data.DataLoader or DataLoaderIter, optional): ...
    """
```

**Comment by AlexGrig** [comment]
Hello,

Thanks for the comments, they sound quite reasonable. I ll implement these changes in the a few days. If something comes up, I ll bring it for discussion here.

**Comment by davidtvs** [comment]
If you notice `TrainLoaderIter` overrides the `__next__` method while `ValLoaderIter` doesn't.

The difference between them is because the user specifies the number of iterations (`num_iter` in `range_test`) he wants for the range test (training) but the validation is always for a full epoch.
So the `__next__` method in `TrainLoaderIter` needs to handle cases where the user specifies `num_iter` larger than the number of iterations in a single epoch of training. On the other hand, `ValLoaderIter` doesn't need to handle that because it always does one full epoch.

**Comment by AlexGrig** [comment]
Ok,  thanks @davidtvs I see now the meaning, I ll implement according to this approach

**Comment by AlexGrig** [comment]
Hello, @davidtvs and @NaleRaphael 
I have finished the planned modifications, if you have more comments feel free to express.

File: torch_lr_finder/lr_finder.py [hunk 1]
```diff
@@ -22,6 +24,52 @@
...
+    def __init__(self, data_loader, auto_reset=True):
+        #super(TrainDataLoaderIter, self).__init__(data_loader)
+        #import pdb; pdb.set_trace()
+        super().__init__(data_loader)
+        self.auto_reset = auto_reset
+        
+    def __next__(self):
+        try:
+            batch = next(self._iterator)
+            inputs, labels, *_ = self.inputs_labels_from_batch(batch)
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:61]: `*_` is not needed

File: torch_lr_finder/lr_finder.py [hunk 2]
```diff
@@ -109,17 +157,22 @@ def range_test(
...
+                However, if a dataset returns different outputs e.g. dicts, then you should inherit from TrainDataLoaderIter
+                class and redefine `inputs_labels_from_batch` method so that it outputs (inputs, labels).     
+            val_loader (torch.utils.data.DataLoader or child of ValDataLoaderIter, optional): if `None` the range test
                 will only use the training loss. When given a data loader, the model is
                 evaluated after each iteration on that dataset and the evaluation loss
                 is used. Note that in this mode the test takes significantly longer but
-                generally produces more precise results. Default: None.
+                generally produces more precise results. 
+                Similarly to `train_loader`, if your dataset outputs are not standard you should inherit 
+                from TrainDataLoaderIter class and redefine method `inputs_labels_from_batch` so that
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:174]: "should inherit from **ValDataLoaderIter** class"

File: torch_lr_finder/lr_finder.py [hunk 3]
```diff
@@ -301,7 +378,7 @@ def _validate(self, dataloader, non_blocking_transfer=True):
         running_loss = 0
         self.model.eval()
         with torch.no_grad():
-            for inputs, labels, *_ in dataloader:
+            for inputs, labels in dataloader:
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:381]: Please rename `dataloader` to `val_iter`

File: torch_lr_finder/lr_finder.py [hunk 4]
```diff
@@ -186,17 +249,31 @@ def range_test(
...
+        else:
+            raise ValueError("`train_loader` has unsupported type: {}".format(type(train_loader)))
+        
+        if val_loader:
+            if isinstance(val_loader, DataLoader):
+                val_iter = ValDataLoaderIter(val_loader)
+            elif isinstance(val_loader, ValDataLoaderIter):
+                val_iter = val_loader
+            else:
+                raise ValueError("`val_loader` has unsupported type: {}".format(type(val_loader)))
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:265]: We should let the user know which types are expected: Dataloader and ValDataLoaderIter

File: torch_lr_finder/lr_finder.py [hunk 5]
```diff
@@ -186,17 +249,31 @@ def range_test(
             raise ValueError("smooth_f is outside the range [0, 1[")
 
         # Create an iterator to get data batch by batch
-        iter_wrapper = DataLoaderIterWrapper(train_loader)
+        if isinstance(train_loader, DataLoader):
+            train_iter = TrainDataLoaderIter(train_loader)
+        elif isinstance(train_loader, TrainDataLoaderIter):
+            train_iter = train_loader
+        else:
+            raise ValueError("`train_loader` has unsupported type: {}".format(type(train_loader)))
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:257]: We should let the user know which types are expected: Dataloader and TrainDataLoaderIter

File: torch_lr_finder/lr_finder.py [hunk 6]
```diff
@@ -22,6 +24,52 @@
...
+        return self
+    
+    def __next__(self):
+        batch = next(self._iterator)
+        return self.inputs_labels_from_batch(batch)
+    
+class TrainDataLoaderIter(DataLoaderIter):
+    def __init__(self, data_loader, auto_reset=True):
+        #super(TrainDataLoaderIter, self).__init__(data_loader)
+        #import pdb; pdb.set_trace()
```
Comment (davidtvs) [inline torch_lr_finder/lr_finder.py:54]: Remove left-over comments

**Comment by davidtvs** [comment]
Merged. Thanks for helping!
